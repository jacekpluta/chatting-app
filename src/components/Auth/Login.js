import React, { Component } from "react";
import {
  Grid,
  Form,
  Segment,
  Button,
  Header,
  Message,
  Icon,
  GridColumn,
} from "semantic-ui-react";
import PropTypes from "prop-types";
import firebase from "../Firebase";
import { Link } from "react-router-dom";
import md5 from "md5";
class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      password: "",
      error: "",
      loading: false,
      registerSuccess: false,
      usersRef: firebase.database().ref("users"),
    };
  }

  componentDidUpdate(prevProps) {
    if (this.props.userRegistered !== prevProps.userRegistered) {
      this.setState({ registerSuccess: true });
    }
  }

  //sets current Form Input row into a state
  handleChange = (event) => {
    this.setState({ [event.target.name]: event.target.value });
  };

  isFormIsValid = () => {
    if (this.state.password.length > 5 || this.state.email.length > 0) {
      return true;
    } else {
      return true;
    }
  };

  handleSubmit = (event) => {
    event.preventDefault();
    if (this.isFormIsValid()) {
      this.setState({ loading: true, error: "" });
      firebase
        .auth()
        .signInWithEmailAndPassword(this.state.email, this.state.password)
        .then((signedInUser) => {
          this.setState({ loading: false });
          this.setState({
            loading: false,
            error: "Invalid login/password",
            email: "",
            password: "",
          });
        })
        .catch((error) => {
          console.log(error);
          this.setState({ loading: false, error: error.message });
        });
    } else {
      this.setState({
        loading: false,
        error: "Invalid login/password",
        email: "",
        password: "",
      });
    }
  };

  getRandomIntInclusive(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  //save user in firebase
  saveUser = (createdUser) => {
    return this.state.usersRef.child(createdUser.user.uid).set({
      name: createdUser.user.displayName,
      avatar: createdUser.user.photoURL,
      email: createdUser.user.email,
      darkmode: false,
      tutorial: true,
      guest: true,
    });
  };

  handleGuestLogin = (event) => {
    event.preventDefault();
    const guestName = `guest${this.getRandomIntInclusive(1, 99999999)}`;
    const guestEmail = `guest${this.getRandomIntInclusive(
      1,
      99999999
    )}@guestmail.com`;
    const guestPassword = `guestPass${this.getRandomIntInclusive(1, 99999999)}`;

    this.setState({ loading: true });
    firebase
      .auth()
      .createUserWithEmailAndPassword(guestEmail, guestPassword)
      .then((createdUser) => {
        // firebase.auth().signOut();
        createdUser.user
          .updateProfile({
            displayName: guestName,
            photoURL: `https://www.gravatar.com/avatar/${md5(
              guestEmail
            )}?d=identicon`,
            darkmode: false,
            tutorial: true,
            guest: true,
          })
          .then(() => {
            this.saveUser(createdUser).then(() => {
              console.log(createdUser);
            });
          })
          .then(() => {
            this.setState({ loading: false, error: "" });
          })
          .catch((err) => {
            console.log("create user error", err);
            this.setState({ loading: false, error: err.message });
          });
      })
      .catch((err) => {
        console.log("create user error", err);
        this.setState({ loading: false, error: err.message });
      });

    // this.setState({ loading: true, error: "" });

    // firebase
    //   .auth()
    //   .signInWithEmailAndPassword(this.state.email, this.state.password)
    //   .then((signedInUser) => {
    //     console.log("singed in", signedInUser);
    //     this.setState({ loading: false });
    //     this.setState({
    //       loading: false,
    //       error: "Invalid login/password",
    //       email: "",
    //       password: "",
    //     });
    //   })
    //   .catch((error) => {
    //     console.log(error);
    //     this.setState({ loading: false, error: error.message });
    //   });
  };

  render() {
    const { email, password } = this.state;
    return (
      <Grid textAlign="center" verticalAlign="middle">
        <GridColumn style={{ maxWidth: 450 }}>
          <Header
            as="h1"
            icon
            color="blue"
            textAlign="center"
            style={{ marginTop: "20px" }}
          >
            <Icon name="chain" color="blue" />
            LOGIN
          </Header>
          <Form size="large" onSubmit={this.handleSubmit}>
            <Segment stacked>
              <Form.Input
                name="email"
                icon="mail"
                iconPosition="left"
                placeholder="Email Adress"
                type="email"
                onChange={this.handleChange}
                value={email}
                className={
                  this.state.error.toLowerCase().includes("email")
                    ? "error"
                    : ""
                }
              />

              <Form.Input
                name="password"
                icon="lock"
                iconPosition="left"
                placeholder="Password"
                type="password"
                onChange={this.handleChange}
                value={password}
                className={
                  this.state.error.toLowerCase().includes("password")
                    ? "error"
                    : ""
                }
              />

              <Button
                disabled={this.state.loading}
                className={this.state.loading ? "loading" : ""}
                color="blue"
                fluid
                size="large"
              >
                SUBMIT
              </Button>
            </Segment>
          </Form>

          {this.state.error !== "" && (
            <Message error>
              {" "}
              <h3>Error</h3>
              {this.state.error}
            </Message>
          )}
          {this.state.registerSuccess && (
            <Message
              success
              header="Your user registration was successful"
              content="You may now log-in with the username you have chosen"
            />
          )}
          <Message>
            Don't have an account?{" "}
            <Link to="/register">
              <Button
                disabled={this.state.loading}
                className={this.state.loading ? "loading" : ""}
                color="red"
                size="mini"
              >
                REGISTER
              </Button>
            </Link>{" "}
            or{" "}
            <Button
              disabled={this.state.loading}
              className={this.state.loading ? "loading" : ""}
              color="green"
              size="mini"
              onClick={this.handleGuestLogin}
            >
              LOG IN AS GUEST
            </Button>
          </Message>
        </GridColumn>
      </Grid>
    );
  }
}

export default Login;

Login.propTypes = {
  userRegistered: PropTypes.bool,
};
