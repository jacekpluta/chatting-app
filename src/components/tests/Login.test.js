import React from "react";

import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import Login from "../Auth/Login";

Enzyme.configure({
  adapter: new Adapter(),
});

let wrapper;

beforeEach(() => {
  wrapper = mount(
    <BrowserRouter>
      <Login />
    </BrowserRouter>
  );
});

afterEach(() => {
  wrapper.unmount();
});

it('should find one button with text "Submit"', () => {
  expect(wrapper.find("button").length).toEqual(1);
  expect(wrapper.find("button").text()).toBe("Submit");
});

it("should find two input fields", () => {
  expect(wrapper.find("input").length).toEqual(2);
});

it("should find one h1 field", () => {
  expect(wrapper.find("h1").length).toEqual(1);
  expect(wrapper.find("h1").text()).toBe("LOGIN");
});

it("should find one form field", () => {
  expect(wrapper.find("form").length).toEqual(1);
});

it("should find input with name email", () => {
  expect(wrapper.find("input").at(0).prop("name")).toEqual("email");
});

it("should find input field email with empty value", () => {
  expect(wrapper.find("input").at(0).prop("value")).toEqual("");
});

it("has a login text area that users can type in", () => {
  wrapper
    .find("input")
    .at(0)
    .simulate("change", { target: { name: "email", value: "test@o2.pl" } });

  wrapper.update();
  expect(wrapper.find("input").at(0).prop("value")).toEqual("test@o2.pl");
});

it("should find input with name password", () => {
  expect(wrapper.find("input").at(1).prop("name")).toEqual("password");
});

it("should find input field password with empty value", () => {
  expect(wrapper.find("input").at(1).prop("value")).toEqual("");
});

it("has a password text area that users can type in", () => {
  wrapper
    .find("input")
    .at(1)
    .simulate("change", { target: { name: "password", value: "test123" } });

  wrapper.update();
  expect(wrapper.find("input").at(1).prop("value")).toEqual("test123");
});

// it('in logs it after typing username, password and clicking button', () => {
// 	wrapper.find('input').at(0).simulate('change', {
// 		target: { name: 'email', value: 'test@o2.pl' }
// 	});
// 	wrapper.find('input').at(1).simulate('change', {
// 		target: { name: 'password', value: 'test123' }
// 	});
// 	expect(wrapper.find('button').text()).toBe('Submit');
// 	wrapper.find('button').simulate('click');
// 	wrapper.find('form').simulate('submit');
// 	wrapper.update();
// });

// it("should render one button object", () => {
//   const wrapper = shallow(<Login />);
//   expect(wrapper.find(<Message/>).length).toEqual(1);
// });

it("renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
