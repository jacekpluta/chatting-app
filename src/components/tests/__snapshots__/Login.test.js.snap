// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ui center aligned middle aligned grid"
>
  <div
    className="column"
    style={
      Object {
        "maxWidth": 450,
      }
    }
  >
    <h1
      className="ui blue icon center aligned header"
      style={
        Object {
          "marginTop": "20px",
        }
      }
    >
      <i
        aria-hidden="true"
        className="blue chain icon"
      />
      Login
    </h1>
    <form
      className="ui large form"
      onSubmit={[Function]}
    >
      <div
        className="ui stacked segment"
      >
        <div
          className="field"
        >
          <div
            className="ui left icon input"
          >
            <input
              name="email"
              onChange={[Function]}
              placeholder="Email Adress"
              type="email"
              value=""
            />
            <i
              aria-hidden="true"
              className="mail icon"
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui left icon input"
          >
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
            <i
              aria-hidden="true"
              className="lock icon"
            />
          </div>
        </div>
        <button
          className="ui blue large fluid button"
          onClick={[Function]}
          role="button"
        >
          Submit
        </button>
      </div>
    </form>
    <div
      className="ui message"
    >
      Don't have an account? 
      <a
        href="/register"
        onClick={[Function]}
      >
        Register
      </a>
    </div>
  </div>
</div>
`;
